### ISSUE TITLE
Validate Transit: make sure all transit stops are part of the roadway network

### ISSUE THREAD

**Issue by e-lo** [wsp-sag/network_wrangler#134]
From Wrangler 1.0...

```python

    def validateWnrsAndPnrs(self):
        """
        Goes through the transit lines in this network and for those that are offstreet (e.g.
        modes 4 or 9), this method will validate that the xfer/pnr/wnr relationships look ship-shape.
        Pretty verbose in the debug log.
        """
        WranglerLogger.debug("Validating Off Street Transit Node Connections")

        nodeInfo = (
            {}
        )  # lineset => { station node => { xfer node => [ walk node, pnr node ] }}
        setToModeType = {}  # lineset => list of ModeTypes ("Local", etc)
        setToOffstreet = {}  # lineset => True if has offstreet nodes
        doneNodes = set()

        critical_found = False

        # For each line
        for line in self:
            if not isinstance(line, TransitLine):
                continue
            # print "validating", line

            lineset = line.name[0:3]
            if lineset not in nodeInfo:
                nodeInfo[lineset] = {}
                setToModeType[lineset] = []
                setToOffstreet[lineset] = False
            if line.getModeType(self.modelType) not in setToModeType[lineset]:
                setToModeType[lineset].append(line.getModeType(self.modelType))
                setToOffstreet[lineset] = setToOffstreet[
                    lineset
                ] or line.hasOffstreetNodes(self.modelType)

            # for each stop
            for stopIdx in range(len(line.n)):
                if not line.n[stopIdx].isStop():
                    continue

                stopNodeStr = line.n[stopIdx].num

                wnrNodes = set()
                pnrNodes = set()

                if stopNodeStr in nodeInfo[lineset]:
                    continue
                nodeInfo[lineset][stopNodeStr] = {}

                # print " check if we have access to an on-street node"
                for link in self.xferli:
                    if not isinstance(link, Linki):
                        continue
                    # This xfer links the node to the on-street network
                    if link.A == stopNodeStr:
                        nodeInfo[lineset][stopNodeStr][link.B] = ["-", "-"]
                    elif link.B == stopNodeStr:
                        nodeInfo[lineset][stopNodeStr][link.A] = ["-", "-"]

                # print " Check for WNR"
                for zac in self.zacs:
                    if not isinstance(zac, ZACLink):
                        continue

                    m = re.match(nodepair_pattern, zac.id)
                    if m.group(1) == stopNodeStr:
                        # this one is invalid for TM1
                        if self.modelType in [Network.MODEL_TYPE_TM1]:
                            errorstr = "ZONEACCESS link should be funnel-stop but stop-funnel found: {}".format(
                                zac
                            )
                            WranglerLogger.critical(errorstr)
                            critical_found = True
                        else:
                            wnrNodes.add(int(m.group(2)))

                    if m.group(2) == stopNodeStr:
                        wnrNodes.add(int(m.group(1)))

                # print "Check for PNR"
                for pnr_filename in self.pnrs.keys():
                    for pnr in self.pnrs[pnr_filename]:
                        if not isinstance(pnr, PNRLink):
                            continue
                        pnr.parseID()
                        if pnr.station == stopNodeStr and pnr.pnr != PNRLink.UNNUMBERED:
                            pnrNodes.add(int(pnr.pnr))

                # print "Check that our access links go from an onstreet xfer to a pnr or to a wnr"
                for link in self.accessli:
                    if not isinstance(link, Linki):
                        continue
                    try:
                        if int(link.A) in wnrNodes:
                            nodeInfo[lineset][stopNodeStr][link.B][0] = link.A
                        elif int(link.B) in wnrNodes:
                            nodeInfo[lineset][stopNodeStr][link.A][0] = link.B
                        elif int(link.A) in pnrNodes:
                            nodeInfo[lineset][stopNodeStr][link.B][1] = link.A
                        elif int(link.B) in pnrNodes:
                            nodeInfo[lineset][stopNodeStr][link.A][1] = link.B
                    except KeyError:
                        # if it's not offstreet then that's ok
                        if not setToOffstreet[lineset]:
                            continue

                        errorstr = (
                            "Invalid access link found in %s lineset %s (incl offstreet) stopNode %s -- Missing xfer?  A=%s B=%s, xfernodes=%s wnrNodes=%s pnrNodes=%s"
                            % (
                                line.getModeType(self.modelType),
                                lineset,
                                stopNodeStr,
                                link.A,
                                link.B,
                                str(nodeInfo[lineset][stopNodeStr].keys()),
                                str(wnrNodes),
                                str(pnrNodes),
                            )
                        )
                        WranglerLogger.warning(errorstr)
                        # raise NetworkException(errorstr)

        nodeNames = {}
        if "CHAMP_node_names" in os.environ:
            book = xlrd.open_workbook(os.environ["CHAMP_node_names"])
            sh = book.sheet_by_index(0)
            for rx in range(0, sh.nrows):  # skip header
                therow = sh.row(rx)
                nodeNames[int(therow[0].value)] = therow[1].value
            # WranglerLogger.info(str(nodeNames))

        # print it all out
        for lineset in nodeInfo.keys():

            stops = nodeInfo[lineset].keys()
            stops.sort()

            WranglerLogger.debug(
                "--------------- Line set %s %s -- hasOffstreet? %s------------------"
                % (lineset, str(setToModeType[lineset]), str(setToOffstreet[lineset]))
            )
            WranglerLogger.debug(
                "%-40s %10s %10s %10s %10s" % ("stopname", "stop", "xfer", "wnr", "pnr")
            )
            for stopNodeStr in stops:
                numWnrs = 0
                stopname = "Unknown stop name"
                if int(stopNodeStr) in nodeNames:
                    stopname = nodeNames[int(stopNodeStr)]
                for xfernode in nodeInfo[lineset][stopNodeStr].keys():
                    WranglerLogger.debug(
                        "%-40s %10s %10s %10s %10s"
                        % (
                            stopname,
                            stopNodeStr,
                            xfernode,
                            nodeInfo[lineset][stopNodeStr][xfernode][0],
                            nodeInfo[lineset][stopNodeStr][xfernode][1],
                        )
                    )
                    if nodeInfo[lineset][stopNodeStr][xfernode][0] != "-":
                        numWnrs += 1

                if numWnrs == 0 and setToOffstreet[lineset]:
                    errorstr = (
                        "Zero wnrNodes or onstreetxfers for stop %s!" % stopNodeStr
                    )
                    WranglerLogger.critical(errorstr)
                    critical_found = True

        if critical_found:
            raise NetworkException("Critical errors found")

```

**Comment by kulshresthaa** [comment]
@e-lo where exactly should these checks be done.. while reading transit network or separate calls on the network after transit network is being read?

`net = TransitNetwork(feed=new_feed, config=updated_config)`

and have the check in constructor? 

```
    def __init__(self, feed: DotDict = None, config: nx.DiGraph = None):
        """
        Constructor
        """
        self.feed: DotDict = feed
        self.config: nx.DiGraph = config
        self.road_net: RoadwayNetwork = None
        self.graph: nx.MultiDiGraph = None

        **if not self.validate_frequencies():
            raise ValueError("Transit lines with non-positive frequencies exist in the network")**
```

or 

`net = TransitNetwork(feed=new_feed, config=updated_config)`
`net.validate_frequencies()`

**Comment by e-lo** [comment]
For the frequency check having it in the constructor sounds good and then having a variable that you set to confirm that check has been completed after each change/alteration.

So

`self.validated_zero_frequency = False`
`self.validated_road_network_consistency = False`
...until the appropriate methods are run which set them to True

For the road node check: Because you  can have a transit network that doesn't have a road network attached to it, let's make it a separate method that isn't automatically done on the read-in.

**Comment by kulshresthaa** [comment]
@e-lo Can you check the IDs I used to compare the transit stops vs roadway nodes [line 153 and 154 here](https://github.com/wsp-sag/network_wrangler/blob/sprint8_ak/network_wrangler/TransitNetwork.py#L153) and transit shapes vs roadway links [line 184 and 185 here](https://github.com/wsp-sag/network_wrangler/blob/sprint8_ak/network_wrangler/TransitNetwork.py#L184)?
Test is indicating MANY missing stops and shape id in the transit network!

**Comment by e-lo** [comment]
I can't look at this right away, but @i-am-sijia might be better to look at this anyhow since she understands the source better.

**Comment by i-am-sijia** [comment]
@kulshresthaa Can you please try with the latest network that I committed in the develop branch:
https://github.com/wsp-sag/network_wrangler/commit/508062a09725961816e30c0f4b811b48b5d9322f

**Comment by kulshresthaa** [comment]
I updated the network with latest version. However there are still MANY missing stops and shapes in the transit network. 

I am using "model_node_id" (`TransitNetwork.FK_STOPS`) from transit network and "model_node_id"(`Roadway.NODE_FOREIGN_KEY`) from roadway network to compare the stops and nodes.

I am using "shape_model_node_id" (`TransitNetwork.FK_SHAPES`) from transit network and "model_link_id"(`Roadway.UNIQUE_LINK_KEY`) from roadway network to compare the shapes and links. 

@i-am-sijia Can you confirm if the IDs from road and transit network used to compare are correct?

**Comment by i-am-sijia** [comment]
I can think of two reasons:

1.  Transit shape ("shape_model_node_id") should be compared to roadway node ID ("model_node_id"), see [PRD](https://docs.google.com/document/d/1OM2tdbjBxbRHBLeyF-eBCfbR_41DhzggeGJQ9gPSZ8g/edit#heading=h.wmeqskhyto8p).

2.  Some transit routes probably have parts that extend outside of the St Paul city, which means there will be nodes and shapes that are not written out in the St Paul roadway network.  I would recommend you try with the full model roadway network, which can be found [here](https://github.com/wsp-sag/client_met_council_network/tree/network_standard/network_standard).

**Comment by kulshresthaa** [comment]
@e-lo can you look into this when you get a chance?

**Comment by kulshresthaa** [comment]
changed the code to compare transit shape ("shape_model_node_id") to roadway nodes ("model_node_id").

Waiting on new transit network (#138) to finalize these transit validation checks.

**Comment by e-lo** [comment]
This works!  Too well...there are still a bunch of stops that aren't part of the network :-) (so far i think it is just a clipping issue).

But...at some point might be good to also make this visual :-) . I'll add as separate task.
